miner/worker.go

```diff
@@ -1201,60 +1211,131 @@ func (w *worker) commit(uncles []*types.Header, interval func(), update bool, st
 	return nil
 }
 
-func (w *worker) findMostProfitableBundle(bundles []types.Transactions, coinbase common.Address, parent *types.Block, header *types.Header) (types.Transactions, *big.Int, *big.Int, uint64) {
-	maxBundlePrice := new(big.Int)
-	maxTotalEth := new(big.Int)
-	var maxTotalGasUsed uint64
-	maxBundle := types.Transactions{}
+type simulatedBundle struct {
+	txs           types.Transactions
+	mevGasPrice   *big.Int
+	ethToCoinbase *big.Int
+	totalGasUsed  uint64
+}
+
+func (w *worker) generateFlashbotsBundle(bundles []types.Transactions, coinbase common.Address, parent *types.Block, header *types.Header) (simulatedBundle, int, error) {
+	simulatedBundles, err := w.simulateBundles(bundles, coinbase, parent, header)
+	if err != nil {
+		return simulatedBundle{}, 0, err
+	}
+
+	sort.SliceStable(simulatedBundles, func(i, j int) bool {
+		return simulatedBundles[j].mevGasPrice.Cmp(simulatedBundles[i].mevGasPrice) < 0
+	})
+
+	return w.mergeBundles(simulatedBundles, parent, header)
+}
+
+func (w *worker) mergeBundles(bundles []simulatedBundle, parent *types.Block, header *types.Header) (simulatedBundle, int, error) {
+	tailGasPrice := big100Gwei // default to 100 Gwei if can't find good price
+	parentTxs := parent.Transactions()
+	if len(parentTxs) > 0 {
+		lastTx := parentTxs[len(parentTxs)-1]
+		if lastTx.GasPrice().Cmp(big100Gwei) > 0 {
+			tailGasPrice = lastTx.GasPrice()
+		}
+	}
+
+	finalBundle := types.Transactions{}
+
+	state, err := w.chain.StateAt(parent.Root())
+	if err != nil {
+		return simulatedBundle{}, 0, err
+	}
+	gasPool := new(core.GasPool).AddGas(header.GasLimit)
+
+	prevState := state
+	prevGasPool := gasPool
+
+	ethToCoinbase := new(big.Int)
+	var totalGasUsed uint64
+
+	count := 0
+	for _, bundle := range bundles {
+		prevState = state.Copy()
+		prevGasPool = new(core.GasPool).AddGas(gasPool.Gas())
+
+		simmed, err := w.computeBundleGas(bundle.txs, parent, header, state, gasPool)
+		if err != nil || simmed.mevGasPrice.Cmp(tailGasPrice) < 0 || simmed.ethToCoinbase.Cmp(new(big.Int)) < 0 {
+			state = prevState
+			gasPool = prevGasPool
+			continue
+		}
+		log.Info("Included bundle", "ethToCoinbase", ethIntToFloat(simmed.ethToCoinbase), "gasUsed", simmed.totalGasUsed, "bundleScore", simmed.mevGasPrice, "bundleLength", len(simmed.txs), "bundleIndex", count)
+		finalBundle = append(finalBundle, bundle.txs...)
+		ethToCoinbase.Add(ethToCoinbase, bundle.ethToCoinbase)
+		totalGasUsed += bundle.totalGasUsed
+		count++
+	}
+
+	if len(finalBundle) == 0 {
+		return simulatedBundle{}, 0, nil
+	}
+
+	return simulatedBundle{
+		txs:           finalBundle,
+		mevGasPrice:   new(big.Int).Div(ethToCoinbase, new(big.Int).SetUint64(totalGasUsed)),
```

Comment:
if the spec separates adjustedGasPrice and mevEquivalentGasPrice then would be better to rename this or change the spec naming